certificate/certificate.go

```diff
@@ -0,0 +1,323 @@
+package certificate
+
+import (
+	cr "crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"github.com/DIMO-Network/edge-network/commands"
+	"github.com/DIMO-Network/edge-network/internal/gateways"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/smallstep/certificates/api"
+	"github.com/smallstep/certificates/authority"
+	"github.com/smallstep/certificates/ca"
+	"go.step.sm/crypto/jose"
+	"go.step.sm/crypto/x509util"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+const generateChallenge = "/auth/web3/generate_challenge"
+const submitChallenge = "/auth/web3/submit_challenge"
+const domain = "http://127.0.0.1:10000"
+const certificatePath = "/opt/autopi/client.pem"
+
+//go:generate mockgen -source certificate.go -destination mocks/certificate_mock.go
+type Signer interface {
+	Sign(req *api.SignRequest) (*api.SignResponse, error)
+}
+
+type Service struct {
+	logger            zerolog.Logger
+	oauthURL          string
+	oauthClientID     string
+	oauthClientSecret string
+	caURL             string
+	caFingerprint     string
+	certificatePath   string
+	stepCa            Signer
+}
+
+func NewCertificateService(logger zerolog.Logger, env gateways.Environment, client Signer) *Service {
+	// set the auth and ca urls based on the environment
+	var authURL string
+	var caURL string
+	var oauthClientID string
+	var oauthClientSecret string
+	var caFingerprint string
+	if env == gateways.Development {
+		authURL = "https://auth.dev.dimo.zone"
+		caURL = "https://ca.dev.dimo.zone"
+		oauthClientID = "step-ca"
+		oauthClientSecret = "KsQ7pruHob6D3NLFQEg9"
```

Comment:
@JamesReate, for now, the client_id and secret are hardcoded in here; I think we need some mechanism to pull the secrets from the cloud if required. I think it would be a good improvement. Overall we need to extract all configuration from the codebase and store it in the file system. Thoughs?